### PR TITLE
Add a guide for code documentation

### PR DESCRIPTION
**Pull request by facundominguez**
These guidelines are the consensus to which we agreed in leapyear some time ago on how to approach documenting a package.

Having this conversation in Tweag could be edifying for individuals and healthy for projects.

Question like the criteria we use to check if documentation is complete, what should be the purpose of documentation, and what isn't worth documenting are on the table.

### CODE REVIEW

**Comment by curiousleo** [comment]
This is great, thanks for writing it up! I have two comments at this point:
- The guide is just called "guide for code documentation", but all the examples and some specific instructions pertain to Haskell only. My suggestion would be to either make this explicitly a guide for Haskell code documentation, or (in my view even better!) a language-agnostic guide to documentation, with examples in several of the languages we use (Haskell, Rust, perhaps others).
- The links to `inline-java` examples don't contain the hash and could easily break when files are moved around, my suggestion would be to include hashes in the links to future-proof them.

**Comment by Profpatsch** [comment]
For documentation in general, there’s always http://jacobian.org/writing/great-documentation/

**Comment by facundominguez** [comment]
> all the examples and some specific instructions pertain to Haskell only.

This is because Haskell is the Language I master and breath most of the time. I'm not against anyone contributing examples in other languages if they want. Could include a disclaimer like that in the document.

> The links to inline-java examples don't contain the hash and could easily break when files are moved around, my suggestion would be to include hashes in the links to future-proof them.

:+1:

> For documentation in general, there’s always http://jacobian.org/writing/great-documentation/

:+1:

File: documentation/code.md [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+Code documentation should be useful to both users and maintainers of a
+package. Documentation must:
+
+- Provide an overview of what problem each package and module solves.
+- Explain the strategy used to solve the problem.
+- Explain the details of the implementation where the intention of the
+  author is not readily apparent from reading the code.
+- Point to examples and material explaining how to use a package.
+- Help find the code responsible for performing a specific task.
```
Comment (aherrmann) [inline documentation/code.md]: I'd rank these above "explain the details of the implementation" and probably also "explain the strategy used to solve the problem" in priority. From a users's perspective I care more about how to use a package than how it is implemented internally.

**Comment by mboes** [comment]
@tweag/developers would be good to have a few people review this. Better yet, to get approvals.

**Comment by feuerbach** [comment]
I had a quick read, these look sensible. But how exactly will the existence of this guide make a difference? Honestly, I doubt that I or most other folks here will remember about this document's existence next time we're creating or documenting a package.

File: documentation/code.md (hunk 1 above)
Comment (facundominguez) [reply documentation/code.md]: I tried to address this.

**Comment by facundominguez** [comment]
For how to remind people, we could make it part of a checklist in PR templates. Also both guides [here](https://github.com/tweag/guides/blob/master/project-management/) talk of code documentation and could link to this guide. There is a [review guide][https://github.com/tweag/guides/tree/master/review] that could link here too, though I feel like that one probably needs a deeper review.

As for how it makes a difference:

* If there is ever an argument of what documentation is missing in a package, this guide could be used or updated to resolve it.
* If we need to improve the documentation of our packages, this guide should help making an assessment. This allows reviewers to share a common idea of what means for documentation to be ready.

**Comment by facundominguez** [comment]
@YPares, @TerrorJack, @mrkkrp, I'm curious whether you have any thoughts on whether these guidelines are helpful or appropriate to your projects.

**Pull request merged**